src/hooks/useOnyx.ts

```diff
@@ -0,0 +1,82 @@
+import {useMemo} from 'react';
+import {useOnyx as originalUseOnyx} from 'react-native-onyx';
+import type {OnyxCollection, OnyxEntry, OnyxKey, OnyxValue, UseOnyxOptions} from 'react-native-onyx';
+import CONST from '@src/CONST';
+import ONYXKEYS from '@src/ONYXKEYS';
+import type {SearchResults} from '@src/types/onyx';
+import useSearchState from './useSearchState';
+
+type OriginalUseOnyx = typeof originalUseOnyx;
+type OriginalUseOnyxReturnType = ReturnType<OriginalUseOnyx>;
+
+const getDataByPath = (data: SearchResults['data'], path: string) => {
+    // Handle prefixed collections
+    for (const collection of Object.values(ONYXKEYS.COLLECTION)) {
+        if (path.startsWith(collection)) {
+            const key = `${collection}${path.slice(collection.length)}`;
+            return data?.[key as keyof typeof data];
+        }
+    }
+
+    // Handle direct keys
+    return data?.[path as keyof typeof data];
+};
+
+// Helper function to get key data from snapshot
+const getKeyData = <TKey extends OnyxKey, TReturnValue>(snapshotData: SearchResults, key: TKey, initialValue?: TReturnValue): TReturnValue => {
+    if (key.endsWith('_')) {
+        // Create object to store matching entries
+        const result: OnyxCollection<TKey> = {};
+        const prefix = key;
+
+        // Get all keys that start with the prefix
+        Object.entries(snapshotData?.data ?? {}).forEach(([dataKey, value]) => {
+            if (!dataKey.startsWith(prefix)) {
+                return;
+            }
+            result[dataKey] = value as OnyxEntry<TKey>;
+        });
+        return (Object.keys(result).length > 0 ? result : initialValue) as TReturnValue;
+    }
+    return (getDataByPath(snapshotData?.data, key) ?? initialValue) as TReturnValue;
+};
+
+/**
+ * Custom hook for accessing and subscribing to Onyx data with search snapshot support
+ */
+const useOnyx: OriginalUseOnyx = (key, options, dependencies) => {
+    const {isOnSearch, hashKey} = useSearchState();
+    const useOnyxOptions = options as UseOnyxOptions<OnyxKey, OnyxValue<OnyxKey>> | undefined;
+    const {selector: selectorProp, ...optionsWithoutSelector} = useOnyxOptions ?? {};
+
+    // Determine if we should use snapshot data based on search state and key
+    const shouldUseSnapshot = isOnSearch && !key.startsWith(ONYXKEYS.COLLECTION.SNAPSHOT) && CONST.SEARCH.SNAPSHOT_ONYX_KEYS.some((snapshotKey) => key.startsWith(snapshotKey));
+
+    // Create selector function that handles both regular and snapshot data
+    const selector = useMemo(() => {
+        if (!selectorProp) {
+            return undefined;
+        }
+        return (data: OnyxValue<OnyxKey> | undefined) => selectorProp(shouldUseSnapshot ? getKeyData(data as SearchResults, key) : data);
+    }, [selectorProp, shouldUseSnapshot, key]);
+
+    const onyxOptions = {...optionsWithoutSelector, selector};
```

Comment:
```suggestion
    const onyxOptions: UseOnyxOptions<OnyxKey, OnyxValue<OnyxKey>> = {...optionsWithoutSelector, selector, allowDynamicKey: true};
```

@wildan-m I believe we should be using `allowDynamicKey` to `true` here, right? To avoid crashes when changing between snapshot and normal keys

Reply:
@fabioh8010 the dynamic error shown when I only add !!hashKey as `isOnSearch` logic, but after change the logic to

```tsx

        const isOnSearch = ((route?.name === SCREENS.SEARCH.CENTRAL_PANE || route?.name === SCREENS.SEARCH.BOTTOM_TAB) && !!hashKey) || isSearchAttachmentModal;

```

and  also filter the switching by `SNAPSHOT_ONYX_KEYS`, the error never shown, do you think we still need to change the `allowDynamicKey` to `true`?

Reply:
@fabioh8010 something related to navigation seems changes and I see the issue re-surface, I'll commit your suggestion for more safety

Reply:
Yes let's set `allowDynamicKey` to `true` for safeness.